worker/worker.go

```diff
@@ -720,30 +720,65 @@ func (w *worker) objectsKeyHandlerDELETE(jc jape.Context) {
 	jc.Check("couldn't delete object", w.bus.DeleteObject(jc.PathParam("key")))
 }
 
-func (w *worker) rhpActiveContractsHandlerGET(jc jape.Context) {
-	busContracts, err := w.bus.ActiveContracts()
-	if jc.Check("failed to fetch contracts from bus", err) != nil {
+func (w *worker) rhpContractsHandler(jc jape.Context) {
+	var req api.ContractsRequests
+	if jc.Decode(&req) != nil {
 		return
 	}
 
+	// fetch all contracts
 	var contracts []api.Contract
-	err = w.withHosts(jc.Request.Context(), busContracts, func(ss []sectorStore) error {
+	err := w.withHosts(jc.Request.Context(), req.Contracts, func(ss []sectorStore) error {
+		var errs []error
 		for i, store := range ss {
-			rev, err := store.(*session).Revision()
-			if err != nil {
-				return err
-			}
-			contracts = append(contracts, api.Contract{
-				ContractMetadata: busContracts[i],
-				Revision:         rev.Revision,
-			})
+			func() {
+				ctx, cancel := context.WithTimeout(jc.Request.Context(), req.HostTimeout)
+				defer cancel()
+
+				rev, err := store.(*session).Revision(ctx)
+				if err != nil {
+					errs = append(errs, err)
+					return
+				}
+				contracts = append(contracts, api.Contract{
+					ContractMetadata: req.Contracts[i],
+					Revision:         rev.Revision,
+				})
+			}()
+		}
+		if err := joinErrors(errs); err != nil {
+			return fmt.Errorf("couldn't retrieve %d contract(s): %w", len(errs), err)
 		}
 		return nil
 	})
-	if jc.Check("failed to fetch contracts", err) != nil {
-		return
+
+	resp := api.ContractsResponse{Contracts: contracts}
+	if err != nil {
+		resp.Error = err.Error()
+	}
+	jc.Encode(resp)
+}
+
+func joinErrors(errs []error) error {
```

Comment:
I don't love that it's in `worker.go` but at the same time I don't have a better spot in mind right now.

Reply:
Luke removed the util and said something along the lines of "we won't need this often and when we need it we can just paste this function at the bottom of the file" - so that's what this is 🤷‍♂️ . We can get rid of this when upgrade to go 1.20.